### DISCUSSION TITLE
Losing java references when mapping std::vector::at std::vector::at aka 'get' @ByRef

### DISCUSSION THREAD

**Discussion by carlostp12** [bytedeco/javacpp#779] [Q&A] [answered]
Hello all. 

Is there a way to map get function  of std::vector get with @ByVal annotation instead of @ByRef?

For example in the preset 
	 https://github.com/bytedeco/javacpp-presets/blob/master/openpose/src/main/java/org/bytedeco/openpose/presets/openpose.java

You can find the line

	infoMap.put(new Info("std::vector<std::array<float,3> >").pointerTypes("Float3Vector").define());

Which maps the vector on to
	https://github.com/bytedeco/javacpp-presets/blob/master/openpose/src/gen/java/org/bytedeco/openpose/Float3Vector.java

It contains the mapped function from std::vector::at as get, like this

	@Index(function = "at") public native @ByRef Float3 get(@Cast("size_t") long i);

My question: is There a way to map this method with @ByVal instead of @ByRef? I mean to get:
	
	@Index(function = "at") public native @ByVal Float3 get(@Cast("size_t") long i);

The reason to do this is because I end up losing java references when using the get method in java side with the @ByRef.

Or may be there exists a way to keep java-references obtained by this get?

**Comment by saudet** [comment, marked as answer]
You could insert a `@ByVal` method with Info.javaText() I guess, but if you're only worried about the life of the reference 
Please try to use PointerScope: http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/

**Reply by carlostp12** [reply]
I tested with the scope and it really worked.

But in this case I cannot apply it because the objects need to be transferred from/to different places and the life cycle of such objects is not clear.
So, I guess I need to map them by using the @ByVal, but when I try to put like this

	mapper.put(new Info("std::vector<MyType>").pointerTypes("MyTypeVector").define())

I obtain

	@Index(function = "at") public native @ByRef MyType get(@Cast("size_t") long i);
	
Instead of

	@Index(function = "at") public native @ByVal MyType get(@Cast("size_t") long i);

I tried several things, the most promising ones:

	infoMap.put(new Info( "std::vector<MyType>::get").javaText("@Index(function = \"at\") public native @Const @ByVal MyType get();").define();
	infoMap.put(new Info( "std::vector<MyType>::get(uint64_t)").javaText("@Index(function = \"at\") public native @Const @ByVal MyType get();").define();
	
But these lines do not produces any change. May be you have any idea how to do it?

**Reply by saudet** [reply]
You should be able to make copies with the copy constructor when you need to, because that's just how C++ works.
